libc/string_catdup.c

```diff
@@ -0,0 +1,813 @@
+	}
+}
+
+
+TEST(string_cat, strcat_ascii)
+{
+	int i;
+	char buf[SIZE_BIGGER_BUF] = { 0 };
+	char ascii[128] = { 0 };
+
+	for (i = 1; i < 128; i++) {
+		ascii[i - 1] = i;
+	}
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, ascii));
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	for (i = strlen(ascii) + 1; i < SIZE_BIGGER_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_extended_ascii)
+{
+	int i;
+	char buf[SIZE_BIGGER_BUF] = { 0 };
+	char ascii[129] = { 0 };
+
+	for (i = 128; i <= 255; i++) {
+		ascii[i - 128] = i;
+	}
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, ascii));
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	for (i = strlen(ascii) + 1; i < SIZE_BIGGER_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_non_ascii)
+{
+	int i;
+	char buf[SIZE_BUF * 2] = NON_ASCII_1;
+	char buf1[] = NON_ASCII_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(NON_ASCII_1 NON_ASCII_2, buf);
+
+	for (i = strlen(NON_ASCII_1 NON_ASCII_2) + 1; i < SIZE_BUF * 2; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+///////////////////////////////////////////////////////////////////////////////////
+
+TEST(string_cat, strncat_basic)
+{
+	int i;
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, strlen(WORD_2) + 1));
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+
+	for (i = strlen(WORD_1 WORD_2); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_zero_n)
+{
+	int i;
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, 0));
+	TEST_ASSERT_EQUAL_STRING(WORD_1, buf);
+
+	for (i = strlen(WORD_1); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_large_n)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-overflow"
+#pragma GCC diagnostic ignored "-Wrestrict"
+
+	int i;
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, SIZE_MAX));
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+
+	for (i = strlen(WORD_1 WORD_2); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_cat, strncat_trimming)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = WORD_2;
+	const char *str = WORD_1 WORD_2;
+	const size_t n = sizeof(buf1) / 2;
+	const int num_elements = sizeof(WORD_1) + n;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, n));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(str, buf, num_elements - 1);
+	TEST_ASSERT_EQUAL_CHAR('\0', buf[num_elements]);
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_cat, strncat_empty_dest)
+{
+	int i;
+	char buf[SIZE_BUF] = "";
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, strlen(WORD_2) + 1));
```

Comment:
We could use sizeof in such cases to simplify those tests. Please look for the similar occurrences.

```suggestion
	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, sizeof(WORD_2)));
```